File: static/java/FormulaTextOps.java
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable;

import java.util.List;
import java.util.stream.Collectors;

/** Text operations shared by text, select, and lookup fields. */
public interface FormulaTextOps extends FormulaField {

  /**
   * Field reference used in string operations (FIND/LEN with LOWER/TRIM). Lookup fields override
   * this to wrap the reference in {@code ARRAYJOIN(field, ", ")} so Airtable can coerce the array
   * to a string. Defaults to the bare {@code {field}} reference.
   */
  default String stringField() {
    return field();
  }

  /**
   * Exact equality.
   *
   * <p>Named {@code eq} rather than {@code equals} because {@code equals(value)} would overload
   * {@code Object.equals} and silently resolve to the wrong method (java-plan-v2 §2.3.2).
   */
  default String eq(String value) {
    return eq(value, true, false);
  }

  default String eq(String value, boolean caseSensitive) {
    return eq(value, caseSensitive, false);
  }

  default String eq(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(field(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return f + "=" + v;
  }

  /** Equals any of the given values. */
  default String equalsAny(List<String> values) {
    return equalsAny(values, true, false);
  }

  default String equalsAny(List<String> values, boolean caseSensitive, boolean trim) {
    return Formulas.or(values.stream().map(v -> eq(v, caseSensitive, trim)).toList());
  }

  /** Not equal. Named {@code neq} for symmetry with {@code eq} (§2.3.2). */
  default String neq(String value) {
    return neq(value, true, false);
  }

  default String neq(String value, boolean caseSensitive) {
    return neq(value, caseSensitive, false);
  }

  default String neq(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(field(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return f + "!=" + v;
  }

  /** Contains substring. */
  default String contains(String value) {
    return contains(value, true, false);
  }

  default String contains(String value, boolean caseSensitive) {
    return contains(value, caseSensitive, false);
  }

  default String contains(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(stringField(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return "FIND(" + v + "," + f + ")>0";
  }

  /** Contains any of the given values. */
  default String containsAny(List<String> values) {
    return containsAny(values, true, false);
  }

  default String containsAny(List<String> values, boolean caseSensitive, boolean trim) {
    return Formulas.or(values.stream().map(v -> contains(v, caseSensitive, trim)).toList());
  }

  /** Contains all of the given values. */
  default String containsAll(List<String> values) {
    return containsAll(values, true, false);
  }

  default String containsAll(List<String> values, boolean caseSensitive, boolean trim) {
    return Formulas.and(values.stream().map(v -> contains(v, caseSensitive, trim)).toList());
  }

  /** Does not contain substring. */
  default String notContains(String value) {
    return notContains(value, true, false);
  }

  default String notContains(String value, boolean caseSensitive, boolean trim) {
    return Formulas.not(contains(value, caseSensitive, trim));
  }

  /** Starts with the given value. */
  default String startsWith(String value) {
    return startsWith(value, true, false);
  }

  default String startsWith(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(stringField(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return "FIND(" + v + "," + f + ")=1";
  }

  /** Does not start with the given value. */
  default String notStartsWith(String value) {
    return notStartsWith(value, true, false);
  }

  default String notStartsWith(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(stringField(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return "FIND(" + v + "," + f + ")!=1";
  }

  /** Ends with the given value. */
  default String endsWith(String value) {
    return endsWith(value, true, false);
  }

  default String endsWith(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(stringField(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return "FIND(" + v + "," + f + ")=LEN(" + f + ")-LEN(" + v + ")+1";
  }

  /** Does not end with the given value. */
  default String notEndsWith(String value) {
    return notEndsWith(value, true, false);
  }

  default String notEndsWith(String value, boolean caseSensitive, boolean trim) {
    String f = Formulas.wrapField(stringField(), caseSensitive, trim);
    String v = Formulas.wrapValueStr(value, caseSensitive, trim);
    return "FIND(" + v + "," + f + ")!=LEN(" + f + ")-LEN(" + v + ")+1";
  }

  /** Phone number equality (strips spaces, dashes, parens, plus, dots). */
  default String phoneEquals(String value) {
    String normalized =
        value
            .chars()
            .filter(c -> " -().+".indexOf(c) < 0)
            .mapToObj(c -> String.valueOf((char) c))
            .collect(Collectors.joining());
    String stripped =
        "SUBSTITUTE(SUBSTITUTE(SUBSTITUTE(SUBSTITUTE(SUBSTITUTE(SUBSTITUTE("
            + field()
            + ",\" \",\"\"),\"-\",\"\"),\"(\",\"\"),\")\",\"\"),\"+\",\"\"),\".\",\"\")";
    return stripped + "=\"" + Formulas.escapeFormulaString(normalized) + "\"";
  }

  /** Regex match. */
  default String regexMatch(String pattern) {
    // Escape for the formula-string context only; regex escapes like \d become
    // \\d in the formula source, which Airtable's parser unescapes back before
    // handing the pattern to the regex engine.
    return "REGEX_MATCH(" + field() + ",\"" + Formulas.escapeFormulaString(pattern) + "\")";
  }
}

File: static/csharp/FormulaTextOps.cs
namespace MyAirtable;

/// <summary>Text operations shared by text, select, and lookup fields.</summary>
public abstract class FormulaTextOps : FormulaField
{
    protected FormulaTextOps(string fieldId)
        : base(fieldId) { }

    /// <summary>
    /// Field reference used in string operations (FIND/LEN with LOWER/TRIM). Lookup fields override
    /// this to wrap the reference in <c>ARRAYJOIN(field, ", ")</c> so Airtable can coerce the array
    /// to a string. Defaults to the bare <c>{field}</c> reference.
    /// </summary>
    protected virtual string StringField() => Field();

    /// <summary>
    /// Exact equality. Named <c>Eq</c> rather than <c>Equals</c> because <c>Equals(value)</c> would
    /// overload <see cref="object.Equals(object)"/> and silently resolve to the wrong method.
    /// </summary>
    public string Eq(string value, bool caseSensitive = true, bool trim = false) =>
        Formulas.WrapField(Field(), caseSensitive, trim)
        + "="
        + Formulas.WrapValueStr(value, caseSensitive, trim);

    /// <summary>Equals any of the given values.</summary>
    public string EqualsAny(
        IEnumerable<string> values,
        bool caseSensitive = true,
        bool trim = false
    ) => Formulas.Or(values.Select(v => Eq(v, caseSensitive, trim)));

    /// <summary>Not equal. Named <c>Neq</c> for symmetry with <c>Eq</c>.</summary>
    public string Neq(string value, bool caseSensitive = true, bool trim = false) =>
        Formulas.WrapField(Field(), caseSensitive, trim)
        + "!="
        + Formulas.WrapValueStr(value, caseSensitive, trim);

    /// <summary>Contains substring.</summary>
    public string Contains(string value, bool caseSensitive = true, bool trim = false) =>
        "FIND("
        + Formulas.WrapValueStr(value, caseSensitive, trim)
        + ","
        + Formulas.WrapField(StringField(), caseSensitive, trim)
        + ")>0";

    /// <summary>Contains any of the given values.</summary>
    public string ContainsAny(
        IEnumerable<string> values,
        bool caseSensitive = true,
        bool trim = false
    ) => Formulas.Or(values.Select(v => Contains(v, caseSensitive, trim)));

    /// <summary>Contains all of the given values.</summary>
    public string ContainsAll(
        IEnumerable<string> values,
        bool caseSensitive = true,
        bool trim = false
    ) => Formulas.And(values.Select(v => Contains(v, caseSensitive, trim)));

    /// <summary>Does not contain substring.</summary>
    public string NotContains(string value, bool caseSensitive = true, bool trim = false) =>
        Formulas.Not(Contains(value, caseSensitive, trim));

    /// <summary>Starts with the given value.</summary>
    public string StartsWith(string value, bool caseSensitive = true, bool trim = false) =>
        "FIND("
        + Formulas.WrapValueStr(value, caseSensitive, trim)
        + ","
        + Formulas.WrapField(StringField(), caseSensitive, trim)
        + ")=1";

    /// <summary>Does not start with the given value.</summary>
    public string NotStartsWith(string value, bool caseSensitive = true, bool trim = false) =>
        "FIND("
        + Formulas.WrapValueStr(value, caseSensitive, trim)
        + ","
        + Formulas.WrapField(StringField(), caseSensitive, trim)
        + ")!=1";

    /// <summary>Ends with the given value.</summary>
    public string EndsWith(string value, bool caseSensitive = true, bool trim = false)
    {
        var f = Formulas.WrapField(StringField(), caseSensitive, trim);
        var v = Formulas.WrapValueStr(value, caseSensitive, trim);
        return "FIND(" + v + "," + f + ")=LEN(" + f + ")-LEN(" + v + ")+1";
    }

    /// <summary>Does not end with the given value.</summary>
    public string NotEndsWith(string value, bool caseSensitive = true, bool trim = false)
    {
        var f = Formulas.WrapField(StringField(), caseSensitive, trim);
        var v = Formulas.WrapValueStr(value, caseSensitive, trim);
        return "FIND(" + v + "," + f + ")!=LEN(" + f + ")-LEN(" + v + ")+1";
    }

    /// <summary>Phone number equality (strips spaces, dashes, parens, plus, dots).</summary>
    public string PhoneEquals(string value)
    {
        var normalized = new string(value.Where(c => " -().+".IndexOf(c) < 0).ToArray());
        var stripped =
            "SUBSTITUTE(SUBSTITUTE(SUBSTITUTE(SUBSTITUTE(SUBSTITUTE(SUBSTITUTE("
            + Field()
            + ",\" \",\"\"),\"-\",\"\"),\"(\",\"\"),\")\",\"\"),\"+\",\"\"),\".\",\"\")";
        return stripped + "=\"" + Formulas.EscapeFormulaString(normalized) + "\"";
    }

    /// <summary>Regex match.</summary>
    public string RegexMatch(string pattern) =>
        // Escape for the formula-string context only; regex escapes like \d become \\d in the
        // formula source, which Airtable's parser unescapes before handing the pattern to the engine.
        "REGEX_MATCH(" + Field() + ",\"" + Formulas.EscapeFormulaString(pattern) + "\")";
}
